core/src/main/java/tc/oc/pgm/modules/StatsMatchModule.java

```diff
@@ -0,0 +1,233 @@
+package tc.oc.pgm.modules;
+
+import java.text.DecimalFormat;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import net.md_5.bungee.api.ChatColor;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.match.event.MatchFinishEvent;
+import tc.oc.pgm.api.match.factory.MatchModuleFactory;
+import tc.oc.pgm.api.module.exception.ModuleLoadException;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.player.PlayerRelation;
+import tc.oc.pgm.api.player.event.MatchPlayerDeathEvent;
+import tc.oc.pgm.events.ListenerScope;
+import tc.oc.pgm.tracker.damage.ProjectileInfo;
+import tc.oc.util.bukkit.component.Component;
+import tc.oc.util.bukkit.component.ComponentUtils;
+import tc.oc.util.bukkit.component.Components;
+import tc.oc.util.bukkit.component.types.PersonalizedText;
+import tc.oc.util.bukkit.component.types.PersonalizedTranslatable;
+import tc.oc.util.bukkit.nms.NMSHacks;
+
+@ListenerScope(MatchScope.RUNNING)
+public class StatsMatchModule implements MatchModule, Listener {
+
+  public static class Factory implements MatchModuleFactory<StatsMatchModule> {
+
+    @Override
+    public StatsMatchModule createMatchModule(Match match) throws ModuleLoadException {
+      return new StatsMatchModule();
+    }
+  }
+
+  public static class playerStats {
+    private int kills = 0;
+    private int deaths = 0;
+    private int killstreak = 0;
+    private int killstreakMax = 0;
+    private int longestBowKill = 0;
+
+    public void onMurder() {
+      kills++;
+      killstreak++;
+      if (killstreak > killstreakMax) killstreakMax = killstreak;
+    }
+
+    public void onDeath() {
+      deaths++;
+      killstreak = 0;
+    }
+
+    public void setLongestBowKill(double distance) {
+      if (distance > longestBowKill) {
+        longestBowKill = Integer.parseInt(new DecimalFormat("#").format(Math.round(distance)));
+      }
+    }
+
+    public Component getBasicStatsMessage() {
+      String KD;
+      if (deaths == 0) {
+        KD = "0";
+      } else {
+        DecimalFormat df = new DecimalFormat("#.##");
+        KD = df.format(kills / deaths);
+      }
+      return new Component(
+          new PersonalizedTranslatable(
+                  "stats.basic",
+                  new PersonalizedText(Integer.toString(kills), ChatColor.GREEN),
+                  new PersonalizedText(Integer.toString(killstreak), ChatColor.GREEN),
+                  new PersonalizedText(Integer.toString(deaths), ChatColor.RED),
+                  new PersonalizedText(KD, ChatColor.GREEN))
+              .render());
+    }
+  }
+
+  public static Map<String, playerStats> allPlayerStats;
```

Comment:
Please make it private.

Reply:
Also, I would persist player `UUID`s, not player names since `UUID`s are unique by definition.

Reply:
But when i enter in `player#getBukkit#getDisplayName` it keeps the colors and everything when displaying the best, does this also work with UUID? I only see a potential error if players use /nick, but it should be prevented to disguise as someone existing anyways.

Reply:
And why private? It is accessed by `StatCommands`

Reply:
> But when i enter in `player#getBukkit#getDisplayName` it keeps the colors and everything when displaying the best, does this also work with UUID? I only see a potential error if players use /nick, but it should be prevented to disguise as someone existing anyways.

Player names are not unique so you shouldn't use them as a key anywhere.

> And why private? It is accessed by StatCommands

https://dzone.com/articles/following-oop-principles-hermetization-in-java

Reply:
Will this: `match.getPlayer(UUID).getBukkit().getDisplayName()` return null if the player with the given UUID is offline?

Reply:
Yes

Reply:
Then how would i get the playername of a player that has the best results but is offline?

Reply:
You need to store that `PlayerStats`.

Reply:
I got it!!

Reply:
Authors are not online but their name is displayed. The player names can be resolved from their UUID, and should be stored in a cache anyways. It's not stats' responsability to know playernames